shell/common/snapshot_controller_impeller.cc

```diff
@@ -21,37 +24,84 @@ sk_sp<DlImage> DoMakeRasterSnapshot(
     SkISize size,
     const std::shared_ptr<impeller::AiksContext>& context) {
   TRACE_EVENT0("flutter", __FUNCTION__);
+  if (!context) {
+    return nullptr;
+  }
+  // Determine render target size.
+  auto max_size = context->GetContext()
+                      ->GetResourceAllocator()
+                      ->GetMaxTextureSizeSupported();
+  double scale_factor_x =
+      static_cast<double>(max_size.width) / static_cast<double>(size.width());
+  double scale_factor_y =
+      static_cast<double>(max_size.height) / static_cast<double>(size.height());
+  double scale_factor = std::min(1.0, std::min(scale_factor_x, scale_factor_y));
+
+  auto render_target_size = impeller::ISize(size.width(), size.height());
+
+  // Scale down the render target size to the max supported by the
+  // GPU if necessary. Exceeding the max would otherwise cause a
+  // null result.
+  if (scale_factor < 1.0) {
+    render_target_size.width *= scale_factor;
+    render_target_size.height *= scale_factor;
+  }
+
+#if EXPERIMENTAL_CANVAS
+  // Do not use the render target cache as the lifecycle of this texture
+  // will outlive a particular frame.
+  impeller::ISize impeller_size = impeller::ISize(size.width(), size.height());
+  impeller::RenderTargetAllocator render_target_allocator =
+      impeller::RenderTargetAllocator(
+          context->GetContext()->GetResourceAllocator());
+  impeller::RenderTarget target;
+  if (context->GetContext()->GetCapabilities()->SupportsOffscreenMSAA()) {
+    target = render_target_allocator.CreateOffscreenMSAA(
+        *context->GetContext(),  // context
+        impeller_size,           // size
+        /*mip_count=*/1,
+        "Picture Snapshot MSAA",  // label
+        impeller::RenderTarget::
```

Comment:
Nit: This could perhaps be dried up using a ternary since only the last argument differs based on `SupportsOffscreenMSAA` and everything else is repeated.

Reply:
Do you mean adding a new CreateOffscreen that differentiates MSAA via an argument?

Reply:
Nothing as involved as all that. Just `context->GetContext()->GetCapabilities()->SupportsOffscreenMSAA() ? kDefaultColorAttachmentConfigMSAA : kDefaultColorAttachmentConfig`

Reply:
derp. Sorry.